crates/client/mempool/src/block_production.rs

```diff
@@ -0,0 +1,159 @@
+        let mut to_process_iter = txs_to_process.into_iter();
+        // This iterator will consume the first part of `to_process_iter`.
+        let consumed_txs_to_process = to_process_iter.by_ref().take(all_results.len());
+
+        for (exec_result, mempool_tx) in Iterator::zip(all_results.into_iter(), consumed_txs_to_process) {
+            match exec_result {
+                Ok(execution_info) => {
+                    // Note: reverted txs also appear as Ok here.
+                    log::debug!("Successful execution of transaction {:?}", mempool_tx.tx_hash());
+
+                    // exec
+                }
+                Err(err) => {
+                    // Internal Server Error: this silently drops the transaction.
+                    log::error!("Unsuccessful execution of transaction {:?}: {err:#}", mempool_tx.tx_hash());
+                }
+            }
+        }
+
+        // This contains the rest of `to_process_iter`.
+        let rest_txs_to_process: Vec<_> = to_process_iter.collect();
+
+        // Add back the unexecuted transactions to the mempool.
+        self.mempool.readd_txs(rest_txs_to_process);
+
+        Ok(())
+    }
+
+    fn produce_block(&mut self) -> Result<(), Error> {
+        // Complete the block with full bouncer capacity.
+        let finish_block = true;
+        self.update_pending_block(finish_block)?;
+
+        // Convert the pending block to a closed block and save to db.
+
+        // Prepare for next block.
+        let mut pending_block = self.mempool.get_or_create_pending_block()?;
+        let bouncer_config = self.backend.chain_config().bouncer_config.clone();
+        self.executor = ExecutionContext::new(Arc::clone(&self.backend), &pending_block.info)?.tx_executor();
+        self.current_pending_tick = 0;
+
+        Ok(())
+    }
+
+    pub async fn block_production_task(&mut self) -> Result<(), anyhow::Error> {
+        let start = tokio::time::Instant::now();
+
+        let mut interval_block_time = tokio::time::interval_at(start, self.backend.chain_config().block_time);
+        interval_block_time.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+        let mut interval_pending_block_update =
+            tokio::time::interval_at(start, self.backend.chain_config().pending_block_update_time);
+        interval_pending_block_update.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+
+        loop {
+            tokio::select! {
+                _ = interval_block_time.tick() => {
+                    if let Err(err) = self.produce_block() {
+                        log::error!("Block production task has errored: {err:#}");
+                    }
+                },
+                _ = interval_pending_block_update.tick() => {
+                    if let Err(err) = self.update_pending_block(false) {
+                        log::error!("Pending block update task has errored: {err:#}");
+                    }
+                },
```

Comment:
I think it is okay to panic and shut down the whole node if one of those task fail.

Reply:
Yeah, i've been wondering about that. I'll see the kind of errors I actually see before choosing either way.